### PR TITLE
Userplex rewrite

### PR DESCRIPTION
**Pull request by ajvb**
Still needs to be tested:

- `authorizedkeys verify`

---------

### CLI

```
$ userplex --help
NAME:
   userplex - userplex - Manage users in various systems based on a LDAP source

USAGE:
   userplex [global options] command [command options] [arguments...]

VERSION:
   v1.0.0

AUTHORS:
   AJ Bahnken <ajvb@mozilla.com>
   Julien Vehent <jvehent@mozilla.com>

COMMANDS:
   aws             Operations within AWS
   authorizedkeys  Operations within authorizedkeys files
   get-person      Get Person from Person API. Useful for finding correct identifier
   help, h         Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --config value, -c value  Path to userplex config file [$USERPLEX_CONFIG_PATH]
   --help, -h                show help
   --version, -v             print the version
```
```
$ userplex aws --help
NAME:
   userplex aws - Operations within AWS

USAGE:
   userplex aws command [command options] [arguments...]

COMMANDS:
   create  Create user
   reset   Reset user credentials
   delete  Delete user
   sync    Run sync operation
   verify  Verify users against Person API. Outputs report, use `sync` to fix discrepancies.

OPTIONS:
   --help, -h  show help
```
```
$ userplex aws create --help
NAME:
   userplex aws create - Create user

USAGE:
   userplex aws create [username]

```

### CODE REVIEW

File: config.yaml [hunk 1]
```diff
@@ -34,114 +21,48 @@ notifications:
...
+  - ldap_username: bkelso
+    local_username: bob
 
-    - ldapuid: tanderson
-      localuid: neo
+  - ldap_username: tanderson
+    local_username: neo
 
-modules:
+# signinurl: "https://myawsaccount.signin.aws.amazon.com/console"
```
Comment (g-k) [inline config.yaml]: signinurl seems useful. Should it be uncommented for use in generated emails or other notifications?

File: notifications/notifications.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// encryptMailBody retrieves the PGP fingerprint of a recipient from ldap, then
+// queries the gpg server to retrieve the public key and encrypts the body with it.
+func EncryptMailBody(conf Config, origBody []byte, person *person_api.Person) ([]byte, error) {
+	keyid := person.GetPGPPublicKeys()[0]
```
Comment (g-k) [inline notifications/notifications.go:87]: Do we need to handle this failing if the person doesn't have a key set?

File: notifications/notifications.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		log.Errorf("Error creating armor encoding writer: %s", err)
+		return nil, err
+	}
+	_, err = w.Write(encbuf.Bytes())
+	if err != nil {
+		log.Errorf("Error writing armor encoded body: %s", err)
+		return nil, err
+	}
+	w.Close()
```
Comment (g-k) [inline notifications/notifications.go]: Can this write fail?

File: notifications/notifications.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return openpgp.Entity{}, fmt.Errorf("keyserver returned non-200 status code %s", resp.Status)
+	}
+	ents, err := openpgp.ReadArmoredKeyRing(resp.Body)
+	if err != nil {
+		return openpgp.Entity{}, fmt.Errorf("could not read entities: %s", err)
+	}
+	return *ents[0], nil
```
Comment (g-k) [inline notifications/notifications.go:147]: Might need nil and len > 0 checks on `ents`

File: main.go [hunk 5]
```diff
@@ -1,172 +1,276 @@
...
-		log.Fatalf("failed to load configuration: %v", err)
+		log.Fatal(err)
 	}
+}
 
-	// just run once
-	if *once || conf.Cron == "disabled" {
-		run(conf)
-		return
+func createModuleSubcommands(module modules.Module) []cli.Command {
```
Comment (g-k) [inline main.go:68]: :100:

File: main.go [hunk 6]
```diff
@@ -1,172 +1,276 @@
...
+				log.Error("You must specify a subcommand")
+				return nil
+			},
+			Subcommands: createModuleSubcommands(&authorizedkeys.AuthorizedKeysModule{}),
+		},
+		{
+			Name:      "get-person",
+			Usage:     "Get Person from Person API. Useful for finding correct identifier",
+			ArgsUsage: "[user-identifier]",
+			Action: func(c *cli.Context) error {
```
Comment (g-k) [inline main.go]: Might be worth pulling this func out of `main()` so we can tell whether it's using any of the main vars via closure and to keep stuff in main at the same abstraction level (i.e. parses conf and CLI args and runs something else)

File: main.go [hunk 7]
```diff
@@ -1,172 +1,276 @@
...
+				log.Error("You must specify a subcommand")
+				return nil
+			},
+			Subcommands: createModuleSubcommands(&aws.AWSModule{}),
+		},
+		{
+			Name:  "authorizedkeys",
+			Usage: "Operations within authorizedkeys files",
+			Action: func(c *cli.Context) error {
```
Comment (g-k) [inline main.go]: nit: missing subcommand func could be factored out

File: modules/authorizedkeys/authorizedkeys.go [hunk 8]
```diff
@@ -1,196 +1,182 @@
...
-		}
-		log.Printf("[info] %d keys written into %q", len(userkeys[uid]), dest)
 	}
-	return
+
+	return badPaths, nil
 }
 
-func interpolate(orig, uid string) string {
+func interpolate(orig, uid string) (string, error) {
```
Comment (g-k) [inline modules/authorizedkeys/authorizedkeys.go:179]: maybe preprocess known vars then pass them to a https://golang.org/pkg/text/template/ str?

File: modules/aws/aws.go [hunk 9]
```diff
@@ -360,56 +177,47 @@ aws_secret_access_key = %s`,
...
+		log.Errorf("aws %q: failed to add user %q to group %q: %v",
+			awsm.config.AccountName, username, group, err)
+		return err
 	}
-	return
+	return nil
 }
 
-func (r *run) removeIamUser(uid string) {
+func (awsm *AWSModule) deleteIamUser(username string) error {
```
Comment (g-k) [inline modules/aws/aws.go:245]: does `ListAccessKeys` kill their active sessions / deactivate their STS tokens?

**Review by g-k — COMMENTED** [review]
Reviewed what's visible in the PR UI and some stuff locally and left comments on the visible files. To review all the new changes in GH, we could fork a branch with one commit removing `vendor/` then make a PR against that branch adding the new stuff then merge that one first.

Overall it'd be good to:

* recover the module readmes
* recover and default to dryrun mode (or add an interactive mode)
* add tests (these could be e2e e.g. given a command, mock server w/ prerecorded and sanitized responses, check the dryrun output is as expected)

+1 for the modules/ breakdown and removing the giant vendored deps.
+1 for switching to the people API which makes things sane

File: modules/aws/aws.go (hunk 9 above)
Comment (ajvb) [reply modules/aws/aws.go:245]: No, but implementing something to do this in userplex is a really interesting idea. Due to how STS tokens work (they can't be revoked, you have to change the permissions of the role they were created for, etc), this isn't really feasible for this initial rewrite, but I love the idea.

File: config.yaml (hunk 1 above)
Comment (ajvb) [reply config.yaml]: Yeah, need to remove this. It's actually created by using the `account_name` like:

```
fmt.Sprintf("https://%s.signin.aws.amazon.com/console", account_name)
```

**Comment by ajvb** [comment]
@g-k thank you for the review! Fixing the stuff you called out.

>     * recover and default to dryrun mode (or add an interactive mode)

The `sync` operation (i.e. `userplex aws sync`) runs in an interactive mode since it can delete many users at once, but for create/delete/reset I felt like it wasn't needed since you must specify the specific user. Do you think it would be good to add it back for the create/reset/delete actions though?

>     * recover the module readmes
>
>     * add tests (these could be e2e e.g. given a command, mock server w/ prerecorded and sanitized responses, check the dryrun output is as expected)

+1, will do.

**Pull request merged**